File: Android/app/src/main/java/com/dkaluta/prosary/reminders/ReminderScheduler.kt
package com.dkaluta.prosary.reminders

import android.Manifest
import android.app.AlarmManager
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.content.ContextCompat
import com.dkaluta.prosary.content.prayerpack.PrayerPackStore
import com.dkaluta.prosary.R
import com.dkaluta.prosary.models.MultiDayRun
import com.dkaluta.prosary.models.MultiDayRuns
import com.dkaluta.prosary.models.Prayer
import com.dkaluta.prosary.models.PrayerKind
import java.util.Calendar

/**
 * Schedules/cancels daily repeating local notifications for a [Prayer]'s saved reminders, via
 * [AlarmManager]. Inexact `setRepeating` (not `setExactAndAllowWhileIdle`) is used deliberately —
 * a prayer reminder doesn't need to-the-second precision, and exact alarms require either a
 * hard-to-qualify special app category or a user trip to Settings on API 31+.
 */
object ReminderScheduler {
    const val NotificationChannelId = "prayer_reminders"
    const val ExtraPrayerId = "prayerId"
    const val ExtraPrayerName = "prayerName"
    // The body is resolved at scheduling time (when the pack manifests are loaded) and carried in
    // the intent, mirroring iOS baking the body into the notification content when arming — the
    // receiver can then run without loading any packs. An alarm armed before this extra existed
    // falls back to the generic body until its next re-arm (boot or edit).
    const val ExtraBody = "body"

    fun createNotificationChannel(context: Context) {
        val manager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        // Idempotent: creating a channel that already exists with the same id is a no-op.
        val channel = NotificationChannel(NotificationChannelId, "Prayer Reminders", NotificationManager.IMPORTANCE_DEFAULT)
        manager.createNotificationChannel(channel)
    }

    fun hasNotificationPermission(context: Context): Boolean {
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.TIRAMISU) return true
        return ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS) ==
            PackageManager.PERMISSION_GRANTED
    }

    /** Replaces all pending alarms for [prayer] with its current enabled reminders. */
    fun schedule(context: Context, prayer: Prayer) {
        cancelAll(context, prayer)
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager

        for (reminder in prayer.reminders) {
            if (!reminder.isEnabled) continue
            val triggerAt = nextTriggerTimeMillis(reminder.hour, reminder.minute)
            val pendingIntent = pendingIntentFor(context, prayer, reminder.id)
            alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, triggerAt, AlarmManager.INTERVAL_DAY, pendingIntent)
        }
    }

    /** Removes all pending alarms for [prayer]. */
    fun cancelAll(context: Context, prayer: Prayer) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        for (reminder in prayer.reminders) {
            alarmManager.cancel(pendingIntentFor(context, prayer, reminder.id))
        }
    }

    /** Re-schedules every enabled reminder across every favorite — used after device reboot,
     * since AlarmManager alarms don't survive it (unlike iOS's UNUserNotificationCenter, which
     * persists at the OS level). */
    fun rescheduleAll(context: Context, prayers: List<Prayer>) {
        for (prayer in prayers) {
            if (prayer.reminders.any { it.isEnabled }) schedule(context, prayer)
        }
    }

    // MARK: Multi-day series

    /** A series in progress earns one alarm per remaining day — the spec's "a notification per
     * day prompting you to continue" — rather than a daily repeat that would keep nagging after
     * the last day. Rewritten from scratch on every call, so recording a day, starting over, or
     * finishing the run all leave exactly the right ones armed. Mirrors iOS's refreshSeries. */
    fun refreshSeries(context: Context, devotionId: String) {
        val alarmManager = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val definition = PrayerPackStore.definition(devotionId)
        val days = definition?.days.orEmpty()

        // Cancel every day's alarm first: which days are owed changes with each session.
        for (day in days.indices) {
            alarmManager.cancel(seriesPendingIntent(context, devotionId, day, days.size))
        }

        if (days.size <= 1 || (definition?.dayProgression ?: "series") != "series") return
        val run = MultiDayRuns.run(context, devotionId) ?: return
        if (run.isComplete(days.size)) return

        val (hour, minute) = reminderTime(definition?.suggestedReminderTime)
        for ((day, triggerAt) in pendingSeriesDays(run, days.size, hour, minute)) {
            alarmManager.set(
                AlarmManager.RTC_WAKEUP, triggerAt,
                seriesPendingIntent(context, devotionId, day, days.size),
            )
        }
    }

    /** Which days still deserve a prompt and when: each unprayed day on the calendar date the
     * run puts it on, skipping anything already past. Pure, so the dates are testable. */
    fun pendingSeriesDays(
        run: MultiDayRun,
        dayCount: Int,
        hour: Int,
        minute: Int,
        now: Long = System.currentTimeMillis(),
    ): List<Pair<Int, Long>> = (0 until dayCount).mapNotNull { day ->
        if (day in run.prayedDays) return@mapNotNull null
        val cal = Calendar.getInstance().apply {
            timeInMillis = run.startedOn
            add(Calendar.DAY_OF_YEAR, day)
            set(Calendar.HOUR_OF_DAY, hour)
            set(Calendar.MINUTE, minute)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        if (cal.timeInMillis > now) day to cal.timeInMillis else null
    }

    /** The bundle's suggested "HH:mm", or early evening — when the day's prayer is traditionally
     * said and, failing that, when someone is most likely free to say it. */
    fun reminderTime(suggested: String?): Pair<Int, Int> {
        val parts = suggested?.split(':')?.mapNotNull { it.toIntOrNull() } ?: emptyList()
        if (parts.size != 2 || parts[0] !in 0..23 || parts[1] !in 0..59) return 18 to 0
        return parts[0] to parts[1]
    }

    private fun seriesPendingIntent(
        context: Context,
        devotionId: String,
        day: Int,
        dayCount: Int,
    ): PendingIntent {
        val name = PrayerPackStore.info(devotionId)?.localizedDisplayName ?: devotionId
        val intent = Intent(context, ReminderBroadcastReceiver::class.java).apply {
            putExtra(ExtraPrayerId, "series:$devotionId:$day")
            putExtra(ExtraPrayerName, name)
            putExtra(
                ExtraBody,
                context.getString(R.string.multi_day_reminder_body, day + 1, dayCount),
            )
        }
        return PendingIntent.getBroadcast(
            context, "series:$devotionId:$day".hashCode(), intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    /** Notification body text per devotion — mirrors iOS's `ReminderScheduler.notificationBody(for:)`:
     * a generic devotion's body comes from its bundle manifest's `reminderBody` (e.g. the
     * Angelus's bell text), not any hardcoded per-kind table. */
    fun notificationBody(prayer: Prayer): String = when (prayer.kind) {
        PrayerKind.Rosary -> "Time to pray the Rosary."
        PrayerKind.JesusPrayer -> "Time for the Jesus Prayer."
        PrayerKind.Custom -> prayer.customDevotionId
            ?.let { PrayerPackStore.info(it)?.localizedReminderBody }
            ?: "Time to pray."
    }

    private fun pendingIntentFor(context: Context, prayer: Prayer, reminderId: String): PendingIntent {
        val intent = Intent(context, ReminderBroadcastReceiver::class.java).apply {
            putExtra(ExtraPrayerId, prayer.id)
            putExtra(ExtraPrayerName, prayer.name)
            putExtra(ExtraBody, notificationBody(prayer))
        }
        val requestCode = (prayer.id + reminderId).hashCode()
        return PendingIntent.getBroadcast(
            context, requestCode, intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE,
        )
    }

    private fun nextTriggerTimeMillis(hour: Int, minute: Int): Long {
        val cal = Calendar.getInstance().apply {
            set(Calendar.HOUR_OF_DAY, hour)
            set(Calendar.MINUTE, minute)
            set(Calendar.SECOND, 0)
            set(Calendar.MILLISECOND, 0)
        }
        if (cal.timeInMillis <= System.currentTimeMillis()) {
            cal.add(Calendar.DAY_OF_YEAR, 1)
        }
        return cal.timeInMillis
    }
}

File: iOS/Prosary/Support/ReminderScheduler.swift

//
//  ReminderScheduler.swift
//  Prosary
//
//  Thin wrapper around UNUserNotificationCenter that schedules/removes daily repeating
//  local notifications for a Prayer's saved reminders.
//

import UserNotifications
import Foundation

struct ReminderScheduler {

  /// Requests notification permission if not yet determined. Returns true when the app
  /// is authorized (already was or just granted).
  @discardableResult
  static func requestPermission() async -> Bool {
    let center = UNUserNotificationCenter.current()
    let settings = await center.notificationSettings()
    switch settings.authorizationStatus {
    case .authorized, .provisional, .ephemeral:
      return true
    case .notDetermined:
      return (try? await center.requestAuthorization(options: [.alert, .sound])) ?? false
    default:
      return false
    }
  }

  /// Replaces all pending notifications for `prayer` with its current enabled reminders.
  static func schedule(for prayer: Prayer) {
    let center = UNUserNotificationCenter.current()
    let prefix = notificationPrefix(for: prayer)

    center.getPendingNotificationRequests { existing in
      // Remove everything that was previously scheduled for this prayer.
      let stale = existing.filter { $0.identifier.hasPrefix(prefix) }.map { $0.identifier }
      center.removePendingNotificationRequests(withIdentifiers: stale)

      for reminder in prayer.reminders where reminder.isEnabled {
        let content = UNMutableNotificationContent()
        content.title = prayer.name
        content.body = notificationBody(for: prayer)
        content.sound = .default
        content.userInfo = ["prayerId": prayer.id.uuidString]

        var comps = DateComponents()
        comps.hour = reminder.hour
        comps.minute = reminder.minute
        let trigger = UNCalendarNotificationTrigger(dateMatching: comps, repeats: true)

        let request = UNNotificationRequest(
          identifier: "\(prefix)\(reminder.id)",
          content: content,
          trigger: trigger
        )
        center.add(request)
      }
    }
  }

  /// Removes all pending notifications for `prayer`.
  static func removeAll(for prayer: Prayer) {
    let center = UNUserNotificationCenter.current()
    let prefix = notificationPrefix(for: prayer)
    center.getPendingNotificationRequests { existing in
      let ids = existing.filter { $0.identifier.hasPrefix(prefix) }.map { $0.identifier }
      center.removePendingNotificationRequests(withIdentifiers: ids)
    }
  }

  // MARK: - Multi-day series

  /// A series in progress earns one notification per remaining day — the spec's "a notification
  /// per day prompting you to continue" — rather than a daily repeat that would keep nagging
  /// after the last day. Rewritten from scratch on every call, so recording a day, starting
  /// over, or finishing the run all leave exactly the right ones pending.
  static func refreshSeries(devotionId: String) {
    let center = UNUserNotificationCenter.current()
    let prefix = seriesPrefix(devotionId)

    center.getPendingNotificationRequests { existing in
      let stale = existing.filter { $0.identifier.hasPrefix(prefix) }.map(\.identifier)
      center.removePendingNotificationRequests(withIdentifiers: stale)
    }

    Task { @MainActor in
      guard let definition = PrayerPackStore.definition(for: devotionId),
            let days = definition.days, days.count > 1,
            (definition.dayProgression ?? .series) == .series,
            let run = MultiDayRuns.run(for: devotionId),
            !run.isComplete(dayCount: days.count) else { return }

      let time = reminderTime(definition.suggestedReminderTime)
      let name = PrayerPackStore.info(for: devotionId)?.localizedDisplayName ?? devotionId
      let pending = Self.pendingSeriesDays(run: run, dayCount: days.count, time: time)
      guard !pending.isEmpty, await requestPermission() else { return }

      for (day, date) in pending {
        let content = UNMutableNotificationContent()
        content.title = name
        content.body = String(
          localized: "multiDay.reminderBody",
          defaultValue: "Day \(day + 1) of \(days.count) awaits.")
        content.sound = .default
        content.userInfo = ["devotionId": devotionId, "dayIndex": day]

        let components = Calendar.current.dateComponents([.year, .month, .day, .hour, .minute], from: date)
        let request = UNNotificationRequest(
          identifier: "\(prefix)\(day)",
          content: content,
          trigger: UNCalendarNotificationTrigger(dateMatching: components, repeats: false))
        try? await center.add(request)
      }
    }
  }

  static func removeSeries(devotionId: String) {
    let center = UNUserNotificationCenter.current()
    let prefix = seriesPrefix(devotionId)
    center.getPendingNotificationRequests { existing in
      let ids = existing.filter { $0.identifier.hasPrefix(prefix) }.map(\.identifier)
      center.removePendingNotificationRequests(withIdentifiers: ids)
    }
  }

  /// Which days still deserve a notification and when: each unprayed day on the calendar date
  /// the run puts it on, skipping anything already past. Pure, so the dates are testable
  /// without a notification centre.
  static func pendingSeriesDays(
    run: MultiDayRun,
    dayCount: Int,
    time: (hour: Int, minute: Int),
    now: Date = Date()
  ) -> [(day: Int, date: Date)] {
    let calendar = Calendar.current
    let start = calendar.startOfDay(for: run.startedOn)
    return (0..<dayCount).compactMap { day in
      guard !run.prayedDays.contains(day),
            let midnight = calendar.date(byAdding: .day, value: day, to: start),
            let fire = calendar.date(bySettingHour: time.hour, minute: time.minute, second: 0, of: midnight),
            fire > now else { return nil }
      return (day, fire)
    }
  }

  /// The bundle's suggested "HH:mm", or early evening — when the day's prayer is traditionally
  /// said and, failing that, when someone is most likely free to say it.
  static func reminderTime(_ suggested: String?) -> (hour: Int, minute: Int) {
    let parts = suggested?.split(separator: ":").compactMap { Int($0) } ?? []
    guard parts.count == 2, (0...23).contains(parts[0]), (0...59).contains(parts[1]) else {
      return (hour: 18, minute: 0)
    }
    return (hour: parts[0], minute: parts[1])
  }

  private static func seriesPrefix(_ devotionId: String) -> String {
    "prosary-series-\(devotionId)-"
  }

  private static func notificationPrefix(for prayer: Prayer) -> String {
    "prosary-\(prayer.id)-"
  }

  private static func notificationBody(for prayer: Prayer) -> String {
    switch prayer.kind {
    case .rosary:      return "Time to pray the Rosary."
    case .jesusPrayer: return "Time for the Jesus Prayer."
    case .custom:
      // Each bundle devotion ships its own notification body in its manifest (e.g. the
      // Angelus's "The Angelus bell is ringing.").
      guard let devotionId = prayer.customDevotionId,
            let body = PrayerPackStore.info(for: devotionId)?.localizedReminderBody else {
        return "Time to pray."
      }
      return body
    }
  }
}
